doc/developer/design/20240609_error_handling.md

```diff
@@ -0,0 +1,413 @@
+But Kafka does not guarantee schema enforcement, so it is common for bad data to
+slip into a Kafka topic.
+
+Because Kafka sources are the most prone to bad data, we propose to pursue a
+Kafka source-specific solution for handling source decoding errors.
+
+### User experience
+
+#### Dead letter queue
+
+We propose to introduce a [dead letter queue][dlq] (DLQ) for Kafka sources.
+
+Intuitively, the DLQ provides a running log of all decoding errors encountered
+by the source. Materialize users can monitor the count of errors in the DLQ and
+set up alerts whenever the count increases. Users can filter on `mz_timestamp`
+and/or `timestamp` to eliminate old errors that are no longe relevant. (Using
+the former column filters by ingestion time, while filtering on the latter
+column filters by production time.)
+
+Concretely, we propose to add a `REDIRECT ERRORS` option to `CREATE SOURCE` that
+specifies the name of a DLQ table in which to emit information about undecodable
+messages:
+
+```sql
+CREATE SOURCE src TO KAFKA CONNECTION kconn ... WITH (
+    REDIRECT ERRORS = dlq
+)
+```
+
+The specified DLQ table (`dlq` in the example above) must not exist before the
+`CREATE SOURCE` command is executed. Materialize will automatically create the
+DLQ table with the following Kafka-specific structure.
+
+ Name           | Type           | Nullable | Description
+----------------|----------------|----------|------------
+ `mz_timestamp` | `mz_timestamp` | No       | The logical timestamp that the message was reclocked to.
+ `error`        | `text`         | No       | The text of the decoding error.
+ `error_code`   | `text`         | No       | The code of the decoding error.
+ `partition`    | `integer`      | No       | The partition of the Kafka message that failed to decode.
+ `offset`       | `bigint`       | No       | The offset of the Kafka message that failed to decode.
+ `timestamp`    | `timestamp`    | No       | The timestamp of the Kafka message that failed to decode.
+ `key`          | `bytea`        | Yes      | The key bytes of the Kafka message that failed to decode.
+ `value`        | `bytea`        | Yes      | The value bytes of the Kafka message that failed to decode.
+
+The DLQ table works like a subsource. It can be queried via `SELECT`, subscribed
+to via `SUBSCRIBE`, and used as the upstream relation for a Kafka sink.
+
+The DLQ table is append only. Allowing users to specify a retention policy on
+the DLQ table is left to future work. Users cannot issue `INSERT`, `UPDATE`, or
+`DELETE` commands against the DLQ.
+
+When a DLQ table is attached to a source, no errors are emitted to the source's
+*errs stream* (see the [error handling section][error-handling-docs] of the
+`compute::render` module).
+
+**Open questions:**
+
+  * Is there a better name for the option than `REDIRECT ERRORS`?
+  * What type gets reported in the system catalog for DLQ tables? Is it really a
```

Comment:
I was thinking the same thing -- it is likely easier to just treat these as another subsource, very similar to the progress subsource, but then we will have to migrate all the catalog objects over later. If we instead try to represent it as a table now, it might require some additional legwork to introduce the appropriate scaffolding but we would avoid the later migration.
Need to look a bit more at the code to decide this, but are there any concerns from a product perspective in either case?

Reply:
> but then we will have to migrate all the catalog objects over later.

Is the concern that we'd break user-created indexes/views that expect the DLQ to be a subsource?

Also, @morsapaes what's the rough timeline for eliminating subsources?

Reply:
Less worried about that, and more worried about the educational impact—just, like, what do we say about this thing in the docs.

Reply:
> Also, @morsapaes what's the rough timeline for eliminating subsources?

Now that this work took priority over the subsource work, the timeline is..."after this epic". The concern is a mix of what Roshan and Nikhil mentioned: do we make this a subsource to be consistent with the _as is_, but make our lives more complicated down the road when we remove the concept of subsources; or do we model it as a table with the _will be_ in mind, but make it awkward to explain to users in light of how other objects currently work.

Since Kafka users will be largely unfamiliar with the concept of a subsource, I'd be inclined towards the latter, if we find it technically feasible and like a good precursor to the (sub)source refactoring work.